enterprise/backend/src/metabase_enterprise/audit_app/permissions.clj

```diff
@@ -1,28 +1,48 @@
 (ns metabase-enterprise.audit-app.permissions
   (:require
-   [metabase-enterprise.audit-db :refer [default-audit-collection
-                                         default-audit-db-id]]
+   [metabase-enterprise.audit-db
+    :refer [default-audit-collection default-audit-db-id]]
+   [metabase.lib.metadata :as lib.metadata]
    [metabase.models.permissions :as perms]
+   [metabase.models.query.permissions :as query-perms]
    [metabase.public-settings.premium-features :refer [defenterprise]]
+   [metabase.query-processor.store :as qp.store]
    [metabase.shared.util.i18n :refer [tru]]
    [toucan2.core :as t2]))
 
 (def audit-db-view-names
   "Used for giving granular permissions into the audit db. Instead of granting permissions to
    all of the audit db, we query the audit db using the names of each view that starts with v_."
-  ["v_users" "v_content" "v_group_members" "v_alerts_subscriptions" "v_activity"])
+  #{"v_users" "v_content" "v_group_members" "v_alerts_subscriptions" "v_activity"})
+
+(defenterprise check-audit-db-permissions
+  "Checks that a given query is not a native query, and only includes table IDs corresponding to the audit views
+  listed above. (These should be the only tables present in the DB anyway, but this is an extra check as a fallback measure)."
+  :feature :audit-app
+  [query]
+  ;; query->source-table-ids returns a set of table IDs or the ::query-perms/native keyword
+  (let [table-ids-or-native-kw (query-perms/query->source-table-ids query)]
+    (qp.store/with-metadata-provider (:database-id query)
+      (doseq [table-id table-ids-or-native-kw]
```

Comment:
Is this potentially calling doseq on a keyword?

Reply:
no it'll always be a set, but the set can include IDs and/or a keyword